xmppserver/src/main/java/org/jivesoftware/openfire/session/LocalIncomingServerSession.java

```diff
@@ -110,11 +110,11 @@ public class LocalIncomingServerSession extends LocalServerSession implements In
      */
     public static LocalIncomingServerSession createSession(String serverName, XMPPPacketReader reader,
             SocketConnection connection, boolean directTLS) throws XmlPullParserException, IOException {
-        return createSession(serverName, reader.getXPPParser(), connection, directTLS);
+        return createSession(serverName, reader.getXPPParser(), connection, directTLS, false);
     }
 
     public static LocalIncomingServerSession createSession(String serverName, XmlPullParser xpp,
-                                                           Connection connection, boolean directTLS) throws XmlPullParserException, IOException {
+                                                           Connection connection, boolean directTLS, boolean doNotSendXMPPStream) throws XmlPullParserException, IOException {
```

Comment:
This doNotSendXMPPStream argument is added, because it's needed for 'recreating' the session object after TLS has completed, to get rid of any stale data in the old session, right? (The open `stream` element has already been sent at that stage).

Instead of trying to juggle with new Stream instances to work our way towards the goal of having a clean session object, would it be more straightforward to instead create a new method on the Session class/interface, that, if invoked, resets that state. It could be named very explicitly `resetState`. With that, there would be no need to try and force a new Session instance. That would avoid the need for error-prone nuances, like this additional argument only-to-be-used-in-a-very-specific-case, and the porting over of the authenticationMethod that was used in the new `transferConnectionToNewSession` method. That in turn would lead to code that's easier to understand and maintain.

There is a case to be made that with a new Session object, there simply is _no way_ that state from an earlier session survives, but I'd argue that having a very explicit 'reset' method, and what I expect the improvement in readability of the code, are advantageous enough to warrent the change.

Reply:
I think a resetState method is harder to implement in this case given the way sessions are registered and unregistered. I think the problem lies in that the current method for starting a session stream is inside of LocalIncomingServerSession (whereas, we use  dedicated stanza handler logic or the NettySessionInitialiser for outbound connections). LocalIncomingServerSession should not be sending XMPP stanzas - this is a hangover from the old-old implementation pre-MINA.

I don't particularly like passing in a boolean to change behaviour of the function, but you could argue it already has a load of side effects already - when createSession is called, the session should be created and returned. The calling object should then choose whether to send the opening stream stanza or not.

Reply:
I think you're right. The sending of data and the session creation should probably be two separate things. This probably holds true for all types of Sessions. This probably warrants its own refactoring process (which likely goes hand-in-hand with OF-2645).

For now, I refactored just the local inbound server-to-server handling, which gets rid of that new argument (the double negation in there really bugged me :) ). Hopefully, this can be a starting point for a larger refactoring (which will go in a different PR).

Reply:
My changes caused failures. I've reverted them for now. The commit is still in, for future reference.